recipes/openai_ab.py

```diff
@@ -0,0 +1,365 @@
+import prodigy.components.preprocess
+import prodigy.util
+import pydantic
+import srsly
+from dotenv import load_dotenv
+from prodigy.util import msg
+
+_ItemT = TypeVar("_ItemT")
+# Set up openai
+load_dotenv()  # take environment variables from .env.
+
+
+class PromptInput(pydantic.BaseModel):
+    id: str
+    prompt_args: Dict[str, Any]
+
+
+class OpenAIPromptAB:
+    display: jinja2.Template
+    prompts: Dict[str, jinja2.Template]
+    inputs: Iterable[PromptInput]
+    batch_size: int
+    verbose: bool
+    randomize: bool
+    openai_api_org: str
+    openai_api_key: str
+    openai_temperature: float
+    openai_max_tokens: int
+    openai_timeout_s: int
+    openai_n: int
+
+    def __init__(
+        self,
+        display: jinja2.Template,
+        prompts: Dict[str, jinja2.Template],
+        inputs: Iterable[PromptInput],
+        *,
+        openai_api_org: str,
+        openai_api_key: str,
+        openai_model: str,
+        batch_size: int = 10,
+        verbose: bool = False,
+        randomize: bool = True,
+        openai_temperature: float = 0,
+        openai_max_tokens: int = 500,
+        openai_timeout_s: int = 50,
+        openai_n: int = 1,
+    ):
+        self.display = display
+        self.inputs = inputs
+        self.prompts = prompts
+        self.model = openai_model
+        self.batch_size = batch_size
+        self.verbose = verbose
+        self.openai_api_org = openai_api_org
+        self.openai_api_key = openai_api_key
+        self.openai_temperature = openai_temperature
+        self.openai_max_tokens = openai_max_tokens
+        self.openai_timeout_s = openai_timeout_s
+        self.openai_n = openai_n
+        self.randomize = randomize
+
+    def __iter__(self) -> Iterable[Dict]:
+        for input_batch in _batch_sequence(self.inputs, self.batch_size):
+            response_batch = self._get_response_batch(input_batch)
+            for input_, responses in zip(input_batch, response_batch):
+                yield self._make_example(
+                    input_.id,
+                    self.display.render(**input_.prompt_args),
+                    responses,
+                    randomize=self.randomize,
+                )
+
+    def progress(self) -> float:
+        return 0.0
+
+    def on_exit(self, ctrl):
+        examples = ctrl.db.get_dataset_examples(ctrl.dataset)
+        counts = Counter()
+        # Get last example per ID
+        for eg in examples:
+            selected = eg.get("accept", [])
+            if not selected or len(selected) != 1 or eg["answer"] != "accept":
+                continue
+            counts[selected[0]] += 1
+        print("")
+        if not counts:
+            msg.warn("No answers found", exits=0)
+        msg.divider("Evaluation results", icon="emoji")
+        pref, _ = counts.most_common(1)[0]
+        msg.good(f"You preferred {pref}")
+        rows = [(name, count) for name, count in counts.most_common()]
+        msg.table(rows, aligns=("l", "r"))
```

Comment:
Ideally this would also show the template that got 0 responses, currently it's:
```
=========================== ✨  Evaluation results ===========================
✔ You preferred prompt2.jinja2

prompt2.jinja2   3
```

Reply:
Fixed! 

```
=========================== ✨  Evaluation results ===========================
✔ It's a draw!

prompt1.jinja2   0
prompt2.jinja2   0
```